### DISCUSSION TITLE
Best way to start a state chart and progressively develop it to a complex state chart

### DISCUSSION THREAD

**Discussion by pke** [statelyai/xstate#2297] [Q&A]
I wonder if there any recommended ways to progressively develop a state chart (using the visualizer).

1. Like first layout the states, that you need. 
2. Define the final state(s)
3. Define actions to transition to states
4. Read actions arguments and transfer them into the context
5. Replace actions with guards
6. Introduce services

Something like that. Are there any tutorials or what is your preferred way (maybe have a video?) @davidkpiano?

**Comment by ChrisShank** [comment]
There are a couple of things that I try to keep in mind when modeling, not sure there is any kind of silver bullet though. What we are trying to model is most likely complicated and trying to think of it all at once will overwhelm you. Embracing the mental models that statecharts provides in order to reduce your cognitive load has been key for me!

- Push off as many implementation detail as much as you can. For me this manifests as externalizing all of the actions, guards and services that you think you machine needs. Dont worry about the implementations of these things yet, the names are the only important part. Focus on describing what that action should do, not how it is going to do it. This also helps you refactor your model early on since model is definitely going to change as you work through it. That said some actions creators like `send` and `log` are pretty declarative and one liners so I typically keep those inline. `assign` is always externalized since it is typically quite imperative.
- Composition abstracts deeper implementations. In `xstate` the main way you can compose is through hierarchal states and invoking services. These forms of composition add “layers” to the behavior you are modeling, which helps reduce how much you have to think through at once. You don't necessarily need to know what is happening *inside* a nested state or an invoked machine yet to reason about the layer you are currently modeling. Once you feel good about the current layer you are modeling then you can move on to model the deeper layers. 
  - This brings up another interesting question, how do you know when to move a hierarchical state into an invoke machine? Its definitely more of an art than a science, but I generally always start with using hierarchical states. There is less overhead and it is easier to refactor when you first start modeling something. I would say that once that nested state starts taking on an *identify* of its own I would recommend putting it into its own machine. This should make it easier to organize (by putting it in its own file), reuse, test. To be honest I’m not entirely sure how to identity when to do this, it is still quite intuitive for me. Maybe someone else could articulate this better?
- Maybe I’m biased here, but I have noticed that `xstate` tends to nudge you to reason from the top down as opposed to the bottom-up (this is further emphasized by the birds-eye view the visualizer gives you!). As a result I always tend to start modeling at the highest “layer” and work myself down. Not sure if others have had the same experienced?
- Embrace the fact that your model is going to change as you work through it! `xstate` forces you to think of edge cases up front and this means that when you are first modeling something it will take a couple of iterations and refactors to get to a good place. Very rarely can I model something ready for production on the first try. Its partly the nature of modeling in my opinion. 

This is a small thing but I have found that I prefer to use '@xstate/inspect` with Vite to model something new as opposed to the visualizer. Right now its a better DX for me. Hopefully the future drag and drop builder changes that!
